### PR TITLE
feat(stat): add cardtime api

### PR DESCRIPTION
**Pull request by hyx930**
### PR types

New features 

### PR changes

APIs

### Describe

feat(stat): add cardtime api

### CODE REVIEW

File: pkg/apiserver/controller/statistics/statistics.go [hunk 1]
```diff
@@ -221,3 +259,218 @@ func convertResultToResponse(response *JobStatisticsResponse, result float64, me
 func checkJobPermission(ctx *logger.RequestContext, job *model.Job) bool {
 	return common.IsRootUser(ctx.UserName) || ctx.UserName == job.UserName
 }
+
+func GetCardTimeByQueueName(ctx *logger.RequestContext, queueName string, startTimeStr string, endTimeStr string) (*GetCardTimeResponse, error) {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: 1. Add function comments
2. rename function name, `GetCardTimeByQueue`

File: pkg/apiserver/router/v1/statistics.go [hunk 2]
```diff
@@ -17,8 +17,10 @@ limitations under the License.
 package v1
 
 import (
+	"github.com/PaddlePaddle/PaddleFlow/pkg/common/logger"
```
Comment (D0m021ng) [inline pkg/apiserver/router/v1/statistics.go]: format import

File: pkg/apiserver/router/v1/statistics.go [hunk 3]
```diff
@@ -39,6 +41,8 @@ func (sr *StatisticsRouter) AddRouter(r chi.Router) {
 
 	r.Get("/statistics/job/{jobID}", sr.getJobStatistics)
 	r.Get("/statistics/jobDetail/{jobID}", sr.getJobDetailStatistics)
+	r.Get("/statistics/cardTime/{queueName}", sr.getCardTimeDetail)
+	r.Post("/stat/job/cardTime", sr.getCardTimeBatch)
```
Comment (D0m021ng) [inline pkg/apiserver/router/v1/statistics.go]: "/statistics/cardTime"

File: pkg/storage/interface.go [hunk 4]
```diff
@@ -17,13 +17,13 @@ limitations under the License.
 package storage
 
 import (
-	log "github.com/sirupsen/logrus"
-	"gorm.io/gorm"
-
 	"github.com/PaddlePaddle/PaddleFlow/pkg/common/logger"
```
Comment (D0m021ng) [inline pkg/storage/interface.go]: format

File: pkg/apiserver/common/utils.go [hunk 5]
```diff
@@ -64,6 +65,8 @@ const (
 
 	FsPrefix = "fs-"
 	UserRoot = "root"
+
+	TIME_LAYOUT = "2006-01-02 15:04:05"
```
Comment (D0m021ng) [inline pkg/apiserver/common/utils.go]: `model.TimeFormat`

File: pkg/storage/interface.go [hunk 6]
```diff
@@ -221,6 +222,7 @@ type JobStoreInterface interface {
 	GetTaskByID(id string) (model.JobTask, error)
 	UpdateTask(task *model.JobTask) error
 	ListTaskByJobID(jobID string) ([]model.JobTask, error)
+	ListJobStatus(startDate, endDate time.Time, queueID, caseType string, minDuration time.Duration) ([]*model.Job, error)
```
Comment (D0m021ng) [inline pkg/storage/interface.go]: `ListJobStat` ?

File: pkg/apiserver/router/v1/statistics.go (hunk 2 above)
Comment (hyx930) [reply pkg/apiserver/router/v1/statistics.go]: ok

File: pkg/apiserver/controller/statistics/statistics_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
+package statistics
+
+import (
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics_test.go:3]: format

File: pkg/storage/job.go [hunk 8]
```diff
@@ -331,3 +331,27 @@ func (js *JobStore) ListTaskByJobID(jobID string) ([]model.JobTask, error) {
 	}
 	return jobList, nil
 }
+
+func (js *JobStore) ListJobStat(startDate, endDate time.Time, queueID, caseType string, minDuration time.Duration) ([]*model.Job, error) {
+	jobStatus := []*model.Job{}
+	var result *gorm.DB
+	switch caseType {
```
Comment (D0m021ng) [inline pkg/storage/job.go]: 这里优化一下逻辑吧

File: pkg/apiserver/controller/statistics/statistics.go [hunk 9]
```diff
@@ -58,6 +69,40 @@ type MetricInfo struct {
...
+	QueueNames []string `json:"queueNames"`
+}
+
+type GetCardTimeResponse struct {
+	QueueName string      `json:"queueName"`
+	CardTime  float64     `json:"cardTime"`
+	Detail    []JobDetail `json:"detail"`
+}
+
+type JobStatusDataForCardTime struct {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: `JobCardTimeInfo`

File: pkg/apiserver/controller/statistics/statistics.go [hunk 10]
```diff
@@ -221,3 +266,232 @@ func convertResultToResponse(response *JobStatisticsResponse, result float64, me
...
+	for resourceKey, resourceValue := range resourceJsonMap {
+		if in(resourceKey, jobResourceList) == true {
+			gpuCards += resourceValue.(int)
+		}
+	}
+	return gpuCards
+}
+
+func FulfillDetailInfo(startTime time.Time, endTime time.Time, detailInfo map[string][]JobStatusDataForCardTime,
+	jobStatusCase []*model.Job, cardTimeCalculation func(*model.Job, time.Time, time.Time, int) float64) (map[string][]JobStatusDataForCardTime, error) {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: `cardTimeCalculation` => calculateCardTimeFunc

File: pkg/apiserver/controller/statistics/statistics_test.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
+package statistics
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics_test.go:1]: Add doc comments

File: pkg/apiserver/controller/statistics/statistics.go [hunk 12]
```diff
@@ -221,3 +254,176 @@ func convertResultToResponse(response *JobStatisticsResponse, result float64, me
...
+		queueID, period, startDate, endDate)
+	if period < minDuration {
+		return nil, 0, fmt.Errorf("time period less than minDuration")
+	}
+	// 初始化detailInfo,map的key为userName，value为[]PaddleJobStatusDataForCardTime
+	limit, offset := 5000, 0
+	//jobStats, err := storage.Job.ListJobStat(startDate, endDate, queueID, minDuration, limit, offset)
+	jobStats, err := storage.Job.ListJobStat(startDate, endDate, queueID, minDuration, limit, offset)
+	if err != nil {
+		logger.Logger().Errorf("[GetCardTimeFromQueueID] list job status for case1 failed, error: %s", err.Error())
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: Return?

File: pkg/apiserver/controller/statistics/statistics.go [hunk 13]
```diff
@@ -221,3 +254,176 @@ func convertResultToResponse(response *JobStatisticsResponse, result float64, me
...
+	period := endDate.Sub(startDate)
+	// 若指定的时间段长度小于最小任务运行时间，报错
+	logger.Logger().Infof("[GetCardTimeFromQueueID]groupID:%v, period:%#v, startDate:%v, endDate:%v",
+		queueID, period, startDate, endDate)
+	if period < minDuration {
+		return nil, 0, fmt.Errorf("time period less than minDuration")
+	}
+	// 初始化detailInfo,map的key为userName，value为[]PaddleJobStatusDataForCardTime
+	limit, offset := 5000, 0
+	//jobStats, err := storage.Job.ListJobStat(startDate, endDate, queueID, minDuration, limit, offset)
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: Delete useless code?

File: pkg/storage/job.go [hunk 14]
```diff
@@ -331,3 +331,56 @@ func (js *JobStore) ListTaskByJobID(jobID string) ([]model.JobTask, error) {
 	}
 	return jobList, nil
 }
+
+func (js *JobStore) ListJobStat(startDate, endDate time.Time, queueID string, minDuration time.Duration, limit, offset int) (map[string][]*model.Job, error) {
+	JobStatMap := make(map[string][]*model.Job)
+	// case1
+	jobStatusForCase1 := []*model.Job{}
+	result := js.db.Table("job").Select("id,queue_id,user_name,activated_at,finished_at").
```
Comment (D0m021ng) [inline pkg/storage/job.go]: 这里select 内容使用一个常量表示吧

File: pkg/storage/job.go [hunk 15]
```diff
@@ -331,3 +335,56 @@ func (js *JobStore) ListTaskByJobID(jobID string) ([]model.JobTask, error) {
...
+		logger.Logger().Errorf("get job status for %v failed, err %v", "case1", result.Error.Error())
+		return nil, result.Error
+	}
+	JobStatMap["case1"] = jobStatusForCase1
+	// case2
+	jobStatusForCase2 := []*model.Job{}
+	result = js.db.Table("job").Select(jobStatSelectColumn).
+		Where(" queue_id = ?", queueID).Where("activated_at <= ? and activated_at != '0000-00-00 00:00:00'", startDate).
+		Where("finished_at <= ? and finished_at > ?", endDate, startDate).
+		Where(" TIMESTAMPDIFF(Second,?,finished_at) > ?", startDate, int(minDuration.Seconds())).
```
Comment (D0m021ng) [inline pkg/storage/job.go]: 这里移到卡时计算里面去

File: pkg/apiserver/controller/statistics/statistics.go [hunk 16]
```diff
@@ -221,3 +254,182 @@ func convertResultToResponse(response *JobStatisticsResponse, result float64, me
...
+			return int(rValue)
+		}
+	}
+	return 0
+}
+
+func GetGpuCards(jobStatus *model.Job) int {
+	members := jobStatus.Members
+	//cardNameList 存放所有的资源类型
+	cardNameList := []string{"nvidia.com/gpu", "_cgpu"}
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: Global variable?

**Review by D0m021ng — APPROVED** [review]

**Review by luoyuedong — APPROVED** [review]

File: pkg/apiserver/controller/statistics/statistics.go [hunk 17]
```diff
@@ -221,3 +258,222 @@ func convertResultToResponse(response *JobStatisticsResponse, result float64, me
...
+//
+// 返回值：
+// []JobDetail: 任务详情列表
+// float64: 任务卡片运行时间总和
+// error: 错误信息，若成功则为nil
+func GetCardTimeByQueueID(startDate time.Time, endDate time.Time,
+	queueID string, minDuration time.Duration) ([]JobDetail, float64, error) {
+	// endDate 原本为 20xx-xx-xx xx:59:59  加一秒
+	endDate = endDate.Add(time.Second)
+	period := endDate.Sub(startDate)
```
Comment (D0m021ng) [inline pkg/apiserver/controller/statistics/statistics.go]: Move to line 285

**Pull request merged**